src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,340 @@
+ *                 there is a successful match, it's added to this list.
+ *   Pool<T> : a list of enumerated terms for each sygus type
+ *
+ *   let k_0 be a fresh skolem of sygus type T
+ *
+ *   Obs<T> += {(k_0, t, null)}
+ *
+ *   Pool<T> = {}
+ *
+ *   while Sol<k_0> == null
+ *     // enumeration phase
+ *     for each T
+ *       s[z] = nextEnum(T_{z})
+ *       if (s[z] is ground)
+ *         if exists (k', t', s') in Obs<T> s.t. |=_X t = toBuiltIn(s[z])
+ *           markSolved(k', s[z])
+ *         else
+ *           let k be a new variable of type : T
+ *           Obs'<T> += (k, toBuiltIn(s[z]), s[z])
+ *       else if no s' in Pool<T> s.t. s' * sigma = s
+ *           pool<T> += s[z]
+ *           for each (k, t, null) in Obs<T>
+ *             Obs' += matchNewObs(k, s[z])
+ *     // match phase
+ *     while Obs' != {}
+ *       Obs'' = {}
+ *       for each (k, t, null) in Obs' // s = null for all triples in Obs'
+ *         Obs<T> += (k, t, null)
+ *         for each s[z] in Pool<T>
+ *           Obs'' += matchNewObs(k, s[z])
+ *       Obs' = Obs''
+ * }
+ *
+ * matchNewObs(k, s[z]) : Obs'
+ * {
+ *   u = rewrite(toBuiltIn(s[z]))
+ *   if match(u, t) == {toBuiltin(z) -> t'}
+ *     if forall t' exists (k'', t'', s'') in Obs<T> s.t. |=_X t'' = t'
+ *       markSolved(k, s{z -> s''})
+ *     else
+ *       let k' be a new variable of type : typeOf(z)
+ *       CandSol<k> += s{z -> k'}
+ *       Obs'<typeOf(z)> += (k', t', null)
+ * }
+ *
+ * markSolved(k, s)
+ * {
+ *   if Sol<k> != null
+ *     return
+ *   if s is not ground
+ *     instantiate s with arbitrary sygus datatype values
+ *   Sol<k> = s
+ *   CandSols<k> += s
+ *   S = {k}      // a stack to process
+ *   while S != {}
+ *     k' = pop(S)
+ *     for all k'' in CandSols keys
+ *       for all s'[k'] in CandSols<k''>
+ *         s'{k' -> Sol<k'>}
+ *         if s' contains free variables not in Obs
+ *           instantiate s with arbitrary sygus datatype values
+ *         if s' is ground
+ *           Sol<k''> = s'
+ *           push(S, k'')
+ * }
+ */
+class SygusReconstruct : public expr::NotifyMatch
+{
+ public:
+  SygusReconstruct(TermDbSygus* tds, SygusStatistics& s);
+
+  /** reconstruct solution
+   *
+   * Returns (if possible) a sygus encoding of a node that is equivalent to sol
+   * and whose syntax matches the grammar corresponding to sygus datatype stn.
+   * The value reconstructed is set to 1 if we successfully return a node,
+   * otherwise it is set to -1.
+   *
+   * For example, given:
+   *   sol = (* 2 x)
+   *   stn = A sygus datatype encoding the grammar
+   *           Start -> (c_PLUS Start Start) | c_x | c_0 | c_1
+   * This method may return (c_PLUS c_x c_x) and set reconstructed to 1.
+   *
+   * @param sol The target term
+   * @param stn The sygus datatype type encoding the syntax restrictions
+   * @param reconstructed The flag to update, indicating the status of the
```

Comment:
Say the meaning of its values.  Also, feel free to remove/modify this argument.  I believe this is based on the old module, but perhaps this argument isn't necessary anymore.

Reply:
I would like for it to remain as we can use it to indicate that a term is not in the grammar (when it's easy to do so) as opposed to reaching the enumeration limit.